src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,488 @@
+                SessionToken = sessionToken
+            };
+        }
+
+        private static string ExtractSessionTokenFromMechanismProperties(IEnumerable<KeyValuePair<string, string>> properties)
+        {
+            string sessionToken = null;
+            if (properties != null)
+            {
+                foreach (var pair in properties)
+                {
+                    switch (pair.Key.ToUpperInvariant())
+                    {
+                        case awsSessionTokenMechanismPropertyKey:
+                            sessionToken = (string)pair.Value;
+                            break;
+                        default:
+                            throw new ArgumentException($"Unknown AWS property '{pair.Key}'.", "properties");
+                    }
+                }
+            }
+
+            return sessionToken;
+        }
+
+        private static async Task<string> GetEC2Response(Uri baseUri)
+        {
+            // Acquire token
+
+            var tokenRequest = new HttpRequestMessage
+            {
+                RequestUri = new Uri(baseUri, "latest/api/token"),
+                Method = HttpMethod.Put,
+            };
+            tokenRequest.Headers.Add("X-aws-ec2-metadata-token-ttl-seconds", "30");
+            var token = await GetHttpContent(tokenRequest, "Failed to acquire EC2 token.").ConfigureAwait(false);
+
+            // Acquire role name
+
+            var roleRequest = new HttpRequestMessage
+            {
+                RequestUri = new Uri(baseUri, "latest/meta-data/iam/security-credentials/"),
+                Method = HttpMethod.Get
+            };
+            roleRequest.Headers.Add("X-aws-ec2-metadata-token", token);
+            var roleName = await GetHttpContent(roleRequest, "Failed to acquire EC2 role name.").ConfigureAwait(false);
+
+            // Acquire credentials
+
+            var credentialsUri = new Uri(baseUri, "latest/meta-data/iam/security-credentials/");
+            var credentialsRequest = new HttpRequestMessage
+            {
+                RequestUri = new Uri(credentialsUri, roleName),
+                Method = HttpMethod.Get
+            };
+            credentialsRequest.Headers.Add("X-aws-ec2-metadata-token", token);
+            var credentials = await GetHttpContent(credentialsRequest, "Failed to acquire EC2 credentials.").ConfigureAwait(false);
+
+            return credentials;
+        }
+
+        private static async Task<string> GetECSResponse(Uri baseUri, string relativeUri)
+        {
+            var credentialsRequest = new HttpRequestMessage
+            {
+                RequestUri = new Uri(baseUri, relativeUri),
+                Method = HttpMethod.Get
+            };
+
+            return await GetHttpContent(credentialsRequest, "Failed to acquire ECS credentials.").ConfigureAwait(false);
+        }
+
+        private static async Task<string> GetHttpContent(HttpRequestMessage request, string exceptionMessage)
```

Comment:
We have a number of similar constructions in the code. You can find it via searching `when (` in the solution. And as I understand this is exactly the reason why `when` construction was added to avoid the same exception handlers for different exceptions.
>Here we can handle all exceptions without breaking stack trace
I don't think that this is a true statement.

If other reviewers are happy with the current code, I will be ok, but I don't see a reason why we need to have 2 the same handles here.